### PR TITLE
Create XeGPU.md

### PR DESCRIPTION
**Pull request by Jianhui-Li**
initial empty file

Please review these guidelines to help with the review process:
- [ ] Have you provided a meaningful PR description?
- [ ] Have you added a test, a reproducer, or a reference to an issue with a reproducer?
- [ ] Have you tested your changes locally for CPU and GPU devices?
- [ ] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?
- [ ] Have you organized your commits logically and ensured each can be built by itself?

### CODE REVIEW

File: docs/rfcs/XeGPU.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
+# RFC for XeGPU Dialect
+
+## Summary
+The XeGPU dialect provides  an abstraction that closely models Xe instructions to support high-perf GEMM code generation. The matrix instructions being processed at this level exactly match the hardware instructions’ semantics including the matrix sizes. The lowering and optimizations built on top of the XeGPU dialect are target-specific.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:4]: I think you can remove "being processed"

File: docs/rfcs/XeTile.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+For each dimension, the tile size must be divisible by `sg_data`, and the size of `sg_layout` multiplying `sg_data` must be divisible by the tile size. The tile is distributed to sg_data in a round-robin fashion. When the tile size is smaller than the size of sg_layout multiplying sg_data, the tile data is wrapped around and continues to be distributed to subgroup threads. For example, for the tile size [128, 128] and sg_data [32, 128], along the second dimension, there is no more data left to assign after the first subgroup, it wraps around and moves to the beginning of the tile and continues the assignment. Instead, for the first dimension, there is more data left after the first round of distribution, so it move to the next subtile and continue the assignement. As a result, the tile would be sliced to four subtiles with size [32,128], with the following mapping: 
+
```
Comment (charithaintc) [inline docs/rfcs/XeTile.md:109]: the size of `sg_layout` multiplying `sg_data` must be divisible by the tile size. 
what do you mean by this?
for A, [32,128]x[2,2] = [64,256] is not divisible by [128,128]
for B, [128, 32]x[2,2] = [256,64] is not divisible by [128, 128]
for C, [32,32]x[2,2] = [64, 64] the reverse is true.

I would prefer to say this,

for the C tile sg_layout x sg_data must evenly divide the C wg level tile size. For A and B this does not have a meaning in my opinion.

File: docs/rfcs/XeGPU.md [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+## Proposal
+XeGPU dialect models a subset of Xe GPU’s ISA. This is the counterpart of NVGPU and AMDGPU dialects, which provide a bridge dialect in the MLIR gradual lowering. XeGPU dialect works with MLIR memref and vector type and complements with Arith/Math/Vector/Memref dialect. XeGPU operations are introduced when there is a special Xe instruction not modeled by LLVM/SPIR-V dialect, for example, like DPAS and 2D block load. In some cases, one XeGPU op may lower to a sequence of instructions for a dedicated and performance-critical function. For example, create_tdesc is mapped to a fixed sequence of instructions to create an address description.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:7]: "complements with" -> "complements"
"like DPAS" -> "DPAS"

File: docs/rfcs/XeGPU.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+|create_nbarrier	| operation ::= XeGPU.create_nbarrier $nbarrier_id, $nbarrier_role attr-dict : uint8_t, type($nbarrier_role) -> type($nbarrier) | %nbarrier = XeGPU.create_nbarrier %nbarrier_id, %nbarrier_role {num_producers = 2, num_consumers = 2} : Uint8_t, nbarrier_role -> !XeGPU.nbarrier |
+|nbarrier_arrive	| operation ::= XeGPU.nbarrier_arrive $nbarrier_id : type($nbarrier) | XeGPU.nbarrier_arrive %nbarrier : !XeGPU.nbarrier |
+|nbarrier_wait	| operation ::= XeGPU.nbarrier_wait $nbarrier_id : type($nbarrier) | XeGPU.nbarrier_wait %nbarrier : !XeGPU.nbarrier |
+|Mfence	| operation ::= XeGPU.mfence attr-dict | XeGPU.mfence {fence_scope = global} |
+|complile-hint	| operation ::= XeGPU.compile_hint attr-dict	| XeGPU.compile_hint {scheduling_barrier} |
+
+The XeGPU dialect supports lowering from [XeTile dialects]{./XeTile.md}, so the tile-based XeTile operation can be further decomposed to multiple XeGPU ops. For example, XeTile.load_tile operation could be lowered to XeGPU’s load_nd or load_gather operations. Compared with the XeTile dialect, the XeGPU dialect works with even smaller matrix sizes, since XeGPU operations map to one hardware instruction in most cases.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:32]: I think we can remove the "so"

File: docs/rfcs/XeTile.md (hunk 2 above)
Comment (charithaintc) [reply docs/rfcs/XeTile.md:109]: Reword,
For each dimension, the tile size must be divisible by `sg_data`

to 

value in each dim of the sg_data must evenly divide the corresponding value in the wg_tile. For example, 128%128 = 0 and 128%32 = 0 for sg_data = [128, 32] and wg_tile = [128, 128].

File: docs/rfcs/XeGPU.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+|create_nbarrier	| operation ::= XeGPU.create_nbarrier $nbarrier_id, $nbarrier_role attr-dict : uint8_t, type($nbarrier_role) -> type($nbarrier) | %nbarrier = XeGPU.create_nbarrier %nbarrier_id, %nbarrier_role {num_producers = 2, num_consumers = 2} : Uint8_t, nbarrier_role -> !XeGPU.nbarrier |
+|nbarrier_arrive	| operation ::= XeGPU.nbarrier_arrive $nbarrier_id : type($nbarrier) | XeGPU.nbarrier_arrive %nbarrier : !XeGPU.nbarrier |
+|nbarrier_wait	| operation ::= XeGPU.nbarrier_wait $nbarrier_id : type($nbarrier) | XeGPU.nbarrier_wait %nbarrier : !XeGPU.nbarrier |
+|Mfence	| operation ::= XeGPU.mfence attr-dict | XeGPU.mfence {fence_scope = global} |
+|complile-hint	| operation ::= XeGPU.compile_hint attr-dict	| XeGPU.compile_hint {scheduling_barrier} |
+
+The XeGPU dialect supports lowering from [XeTile dialects]{./XeTile.md}, so the tile-based XeTile operation can be further decomposed to multiple XeGPU ops. For example, XeTile.load_tile operation could be lowered to XeGPU’s load_nd or load_gather operations. Compared with the XeTile dialect, the XeGPU dialect works with even smaller matrix sizes, since XeGPU operations map to one hardware instruction in most cases.  
+
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:33]: "Both need to create a tensor descriptor to describe the addresses/offsets to the tensor data, use it to load/store/prefetch, and then update it to the next data blocks."
Seems a bit complex. How about,
"Both need a tensor descriptor to describe the addresses/offsets to a data block. The descriptor is used for load/store/prefetch, and then updated for reuse with the next data block"

File: docs/rfcs/XeGPU.md [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+XeGPU supports two flavors of load/store operations: n-dimension load (nd load) and scattered load. Both need to create a tensor descriptor to describe the addresses/offsets to the tensor data, use it to load/store/prefetch, and then update it to the next data blocks. Nd_load can be used to map to PVC’s 1D load, 2D load, or future nd load. Scattered load requires a special tensor descriptor, which contains one separate address offset for each WI thread.  
+
+`create_nd_tdesc` creates a tensor descriptor for an n-dimensional tensor, which describes a subview of an n-dimensional base tensor. The information of the base tensor is passed as operands including base address, offsets, and strides. The shape and element data type of the tensor view (subtensor) are specified in the output tensor_desc data type, and they must be known at the compile time. The tensor_desc design is extensible for future Xe hardware to support higher-dimension tensors. To create an n-dimension tensor descriptor, the user needs to pass “n” number of base_shape and base_stride for the base nd-tile, and “n” number of offsets, and the shape in the result tensor_desc has to be “n” numbers.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:36]: "To create an n-dimension tensor descriptor, the user needs to pass" ->
"n-dimension tensor descriptor constructor requires"

File: docs/rfcs/XeGPU.md [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+`create_nd_tdesc` creates a tensor descriptor for an n-dimensional tensor, which describes a subview of an n-dimensional base tensor. The information of the base tensor is passed as operands including base address, offsets, and strides. The shape and element data type of the tensor view (subtensor) are specified in the output tensor_desc data type, and they must be known at the compile time. The tensor_desc design is extensible for future Xe hardware to support higher-dimension tensors. To create an n-dimension tensor descriptor, the user needs to pass “n” number of base_shape and base_stride for the base nd-tile, and “n” number of offsets, and the shape in the result tensor_desc has to be “n” numbers.  
+
+The example below creates a 2D tensor_desc with base matrix address, shapes, strides, and the offsets of the 2D subtensor. the tensor_desc “remembers” the base tensor buffer’s information, so when it is used to load the subtensor, the lowering will handle the out-of-boundary access implicitly and preferably using hardware auto-padding features for the out-of-boundary elements. On PVC, the stride of the innermost dimension (base_stride[0]) must be 1.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:38]: "the lowering" -> "lowering"

File: docs/rfcs/XeGPU.md [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+ #sg_map_a = xegpu.sg_map<wi_layout = [2, 8], wi_data = [1, 2]>
+%tdesc1 = XeGPU.create_nd_tdesc %mem_addr, %offsets:2, %base_shape:2,%base_stride:2
+		: uint64, index, index, index, index, index, index
+     	into tensor_desc<8x16xbf16, #sg_map_a>
+
+%tdesc2 = XeGPU.create_nd_tdesc %mem_addr, %offsets:2, %base_shape:2,%base_stride:2 {mode =vc}
+		: uint64, index, index, index, index, index, index
+     	into tensor_desc<8x16xbf16>
+```
+Attribute `xegpu.sg_map` describes the mapping between WI threads and the 2D subtensor specified by the tensor descriptor. With the sg_map, `wi_layout` specifies the layout in which WI threads correspond to the memory, and `wi_data` describes the data block accessed by each WI thread. In the example above, wi_layout=[2, 8] means that each subgroup has 16 WI threads in 2 rows and 8 columns, and wi_data=[1,2] means that each WI thread owns a [1,2] data fragment. The data elements with each data fragment assigned to a WI thread must be contiguous. So the sg_map describes a total [2,16] submatrix at the subgroup level.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:50]: "WI" -> "work item (WI)"

File: docs/rfcs/XeTile.md (hunk 2 above)
Comment (charithaintc) [reply docs/rfcs/XeTile.md:109]: Data distribution among the subgroups work as follows. If sg_data[i] x sg_layout[i] < wg_tile[i] the we have enough data to distribute to the next subgroup along dimension i. So the next slice is assigned to (i+1)'th subgroup in that dimension. If sg_data[i] x sg_layout[i] >= wg_tile[i] we have already used up all the data along that dimension. In that case we wrap around the tile and assign the same tile to (i+1)'th subgroup. For example,....

File: docs/rfcs/XeGPU.md [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+The size of the 2D subtensor referred by the result tensor_desc must be divisible by sg_map size, which comes from wi_layout multiplying with wi_data. For each dimension, the size of the 2D subtensor must be divisible by wi_layout x wi_data. The 2D subtensor size must be larger than or equal to the sg_map size. When the 2D subtensor size is larger than the sg_map size, it is distributed to WI threads in a round-robin fashion.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:52]: "multiplying with" -> "multiplied by"

File: docs/rfcs/XeGPU.md [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+ %tdesc1 = XeGPU.create_nd_tdesc %mref, %offsets:2
+		: memref<1024x1024xbf16>, index, index
+     	into tensor_desc<8x16xbf16, #sg_map_a>
+
+ %tdesc2 = XeGPU.create_nd_tdesc  %mref, %offsets:2 {mode =vc}
+		: memref<1024x1024xbf16>, index, index
+     	into tensor_desc<8x16xbf16>
+```
+
+The example below accepts a memory address and an offset and creates a 1D tensor_desc. The tensor_desc describes a 1D vector that is loaded by all work item (WI) threads combined within the subgroup.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:70]: "work item (WI)" -> "WI" (Assuming acronym is  explained on first usage above)

File: docs/rfcs/XeGPU.md [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+```mlir
+  #sg_map_a = xegpu.sg_map<wi_layout = [2, 8], wi_data = [1, 2]>
+  %result = XeGPU.load_nd %tdesc1 {L1_hint = uncached, L3_hint = uncached} :
+          tensor_desc<8x16xbf16, #sg_map_a> into vector<4x2xbf16>
+
+  %result = XeGPU.load_nd %tdesc2 {L1_hint = uncached, L3_hint = uncached, mode = vc} :
+          tensor_desc<8x16xbf16> into vector<8x16xbf16>
+```
+Attributes `L1_hint`, `L2_hint`, and `L3_hint` can be applied to Load_nd, specifying hint directives for different levels of cache hierarchy. On PVC, cache directive for load could be "uncached, cached, streaming, read_invaldiate".  Streaming means that the data is cached but is more likely to be swapped out, and read_invaldiate simply invalidates the cache line after read. For write, cache policy could be "uncached, write_through, write_back, streaming". Write_through writes to the next level cache immediately, and write_back holds the modification until the cache line is kicked out due to the cache replacement policy.  PVC uses L1_hint and L3_hint and omits L2_hint.  There are only a few valid combinations between L1_hint and L3_hint for PVC.
```
Comment (silee2) [inline docs/rfcs/XeGPU.md:95]: "Load_nd, specifying" -> "Load_nd. They serve as"

**Review by silee2 — APPROVED** [review]
LGTM

**Pull request merged**